### PR TITLE
Add some documentation about what `as` casts are legal

### PR DESCRIPTION
**Pull request by sourcefrog**
Based on the description in https://github.com/rust-lang/rust/blob/219eca11b044de3644b3e9101124513c1a842b09/src/librustc_typeck/check/cast.rs#L11 and https://doc.rust-lang.org/nightly/nomicon/casts.html

### CODE REVIEW

**Comment by sourcefrog** [comment]
I don't feel this is really clear enough at the level of the Rust book, yet, but it's a start.

There doesn't seem to be any definition of the concept of coercion yet.

**Comment by sourcefrog** [comment]
Thanks to Adrian in http://stackoverflow.com/a/33884290/243712 for pointing this out.

**Comment by Gankro** [comment]
Note that https://doc.rust-lang.org/nomicon/casts.html and https://doc.rust-lang.org/nomicon/coercions.html exist. I'm honestly a bit dubious of their existence, though. They were more motivated when I was trying to make the nomicon totally stand-alone and/or a substitute for the reference. Migrating all this content to The Book proper seems like a reasonable choice.

**Comment by steveklabnik** [comment]
Part of the problem with this is that it introduces a forward reference to raw pointers, but, I'm planning on a reorganization of the book anyway, so having this text around can still be useful.

Thanks for adding this!

@bors: r+

**Comment by steveklabnik** [comment]
(@gankro we should chat about your nomicon feels sometime too)

**Comment by sourcefrog** [comment]
Thanks for the speedy review. I'd like to clear up the phrasing a bit before this is actually merged.

**Comment by steveklabnik** [comment]
@bors: r-


What parts in particular?

**Comment by sourcefrog** [comment]
Do you think it's clear enough? 

It refers to some constants like `Sized` and *C-like enum* that aren't explained by this point. For `Sized` I could just point forward to the chapter on traits, and C-like enums are just enums whose values are all integers?

I think it's also possible to handle coercion by defining `Deref`?

**Comment by sourcefrog** [comment]
OK, I think this is clearer. I copied some text about the semantics of numeric casts from the nomicon, which seemed appropriate for an introduction.

If you think this is clear enough and correct, please go ahead and merge.

**Comment by steveklabnik** [comment]
> It refers to some constants like Sized and C-like enum that aren't explained by this point. 

Since this is getting re-arranged in the near future, I'm not as worried about it as I normally would be.

**Comment by steveklabnik** [comment]
I do still think this is great, thanks :+1: 

@bors: r+ rollup

File: src/doc/book/casting-between-types.md [hunk 1]
```diff
@@ -15,7 +37,95 @@ let x: i32 = 5;
...
+
+## Explicit coercions
+
+A cast `e as U` is valid if `e` has type `T` and `T` *coerces* to `U`.
+
+For example:
+
+```rust
+let a = "hello";
+let b = a as String;
```
Comment (Ms2ger) [inline src/doc/book/casting-between-types.md]: ... this doesn't compile.

Comment (sourcefrog) [inline src/doc/book/casting-between-types.md]: :/

I'll send a follow-up

**Comment by sourcefrog** [comment]
Is it considered a bug in rustbook, or in the rust tree's use of it, that it doesn't test the examples? I did run the tests on this and I guess bors did too.

**Comment by steveklabnik** [comment]
It should have. I'm not sure why this passed.

**Comment by steveklabnik** [comment]
I think it's a bug, I'm fixing it. Ugh. :(

**Pull request merged**